Extend kernel-performance-tests to run in containers

dylan-matthew-turner opened the pull request:
## Reason

Customers want to use docker from nilrt, and we need to verify there is not a significant performance cost from doing so, so extend the existing kernel-performance-tests to run from containers to verify docker support is functional.

## Implementation

- Create two containers: one for running cyclictest and one to run in parallel with a load from hackbench, fio, or iperf.
- Duplicate the existing tests and modify each so the load command is called in the parallel container and the cyclictest is called in the main container

## Testing

Ran all the way through on a cRIO 9037 and uploaded results to Grafana.

NOTE: One small hiccup which is `fio` segfaults from a container. This will affect results as no load runs during the fio_containerized_test, but it doesn't stop the test or cause it to fail, so we can likely proceed for now and file a bug to fix the segfault.

amstewart commented on recipes-kernel/kernel-tests/kernel-performance-tests-files/cyclictest-container/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+    && echo " > /var/local/ptest-results/kernel-performance-tests/\${1}-run_cyclictest.log 2>&1" \
+        >> call_run_ct.sh \
+    && echo "echo \"\${CYCLICTEST_RESULT}\"" >> call_run_ct.sh \
+    && chmod +x call_run_ct.sh
```
You can just put this logic into a `call_run_ct.sh` script in the docker context directory, and then install it like
```dockerfile
ADD --chmod=0775 call_run_ct.sh /call_run_ct.sh
```

amstewart commented on recipes-kernel/kernel-tests/kernel-performance-tests-files/parallel-container/Dockerfile:
```diff
@@ -0,0 +1,45 @@
+    iperf3 \
+    psmisc
+
+# Scripts to start loads from outside container
```
Same with all these scripts.

amstewart commented on recipes-kernel/kernel-tests/kernel-performance-tests-files/parallel-container/Dockerfile:
```diff
@@ -0,0 +1,45 @@
+FROM ubuntu:22.04
+
+# Install rt-tests, sudo, etc
+RUN apt update
+RUN DEBIAN_FRONTEND=noninteractive apt install -y \
```
Docker caches the container at each step. So index updates and install operations are normally done as a single command. eg.
```dockerfile
RUN apt update && apt install -y ...
```

dylan-matthew-turner commented:
Summary of changes:

- Move apt update into same command as apt install
- Extract docker script files to separate files instead of constructing via echo
- Adjust calling scripts to use new files

amstewart commented on recipes-kernel/kernel-tests/kernel-performance-tests.bb:
```diff
@@ -11,7 +11,7 @@ FILESEXTRAPATHS:prepend := "${THISDIR}/${PN}-files:"
 
 DEPENDS = "virtual/kernel"
-RDEPENDS:${PN}-ptest += "bash rt-tests fio iperf3 python3 python3-pip"
+RDEPENDS:${PN}-ptest += "bash rt-tests fio iperf3 python3 python3-pip docker"
```
The package name is `docker-ce`.

gratian commented on recipes-kernel/kernel-tests/kernel-performance-tests-files/cyclictest-container/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM ubuntu:22.04
```
Can we use a "latest" tag here instead of "22.04"? (re: https://hub.docker.com/_/ubuntu)
Trying to avoid having to manually rev this version in the future.

gratian commented on recipes-kernel/kernel-tests/kernel-performance-tests-files/parallel-container/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+FROM ubuntu:22.04
```
Same comment as above, is "latest" an option instead of a specific version?

gratian commented on recipes-kernel/kernel-tests/kernel-performance-tests.bb:
```diff
@@ -27,6 +27,16 @@ SRC_URI += "\
     file://test_kernel_cyclictest_hackbench.sh \
     file://test_kernel_cyclictest_fio.sh \
     file://test_kernel_cyclictest_iperf.sh \
+    file://test_kernel_cyclictest_idle_containerized.sh \
```
Given these tests take a while to run (8 hours per test x 4 tests = 32 hours) can you split the containerized versions into a separate recipe? Maybe "kernel-performance-tests-containerized"  or something like that.
It will require separate ATS event/run but it will give us the option to run it in parallel (contingent on ATS HW availability etc.).

dylan-matthew-turner replied:
Can do

dylan-matthew-turner commented:
Changes: moved to new test (forgot to update to ubuntu-latest). More changes coming

dylan-matthew-turner committed:
k-container-prf-tsts: extnd prf tests w containers

    Duplicate existing kernel-performance-tests but now with containers.

    Includes:
    - Dockerfiles for load and main cyclictest
    - Modified version of each of the four perf tests to call their scripts
      from the docker containers

Signed-off-by: Dylan Turner <dylan.turner@ni.com>

dylan-matthew-turner commented:
Changes:
- Moved to new test
- ubuntu:22.04 => ubuntu:latest
